include/nbl/builtin/hlsl/subgroup2/arithmetic_portability_impl.hlsl

```diff
@@ -0,0 +1,228 @@
+            retval[i] = value[i-1];
+        return retval;
+    }
+};
+
+template<class Params, class BinOp, uint32_t ItemsPerInvocation, bool native>
+struct reduction
+{
+    using type_t = typename Params::type_t;
+    using scalar_t = typename Params::scalar_t;
+    using binop_t = typename Params::binop_t;
+    using op_t = reduction<Params, binop_t, 1, native>;
+
+    // NBL_CONSTEXPR_STATIC_INLINE uint32_t ItemsPerInvocation = vector_traits<T>::Dimension;
+
+    scalar_t operator()(NBL_CONST_REF_ARG(type_t) value)
+    {
+        binop_t binop;
+        op_t op;
+        scalar_t retval = value[0];
+        [unroll]
+        for (uint32_t i = 1; i < ItemsPerInvocation; i++)
+            retval = binop(retval, value[i]);
+        return op(retval);
+    }
+};
+
+
+// specs for N=1 uses subgroup funcs
+// specialize native
+#define SPECIALIZE(NAME,BINOP,SUBGROUP_OP) template<class Params, typename T> struct NAME<Params,BINOP<T>,1,true> \
+{ \
+    using type_t = T; \
+ \
+    type_t operator()(NBL_CONST_REF_ARG(type_t) v) {return glsl::subgroup##SUBGROUP_OP<type_t>(v);} \
+}
+
+#define SPECIALIZE_ALL(BINOP,SUBGROUP_OP) SPECIALIZE(reduction,BINOP,SUBGROUP_OP); \
+    SPECIALIZE(inclusive_scan,BINOP,Inclusive##SUBGROUP_OP); \
+    SPECIALIZE(exclusive_scan,BINOP,Exclusive##SUBGROUP_OP);
+
+SPECIALIZE_ALL(bit_and,And);
+SPECIALIZE_ALL(bit_or,Or);
+SPECIALIZE_ALL(bit_xor,Xor);
+
+SPECIALIZE_ALL(plus,Add);
+SPECIALIZE_ALL(multiplies,Mul);
+
+SPECIALIZE_ALL(minimum,Min);
+SPECIALIZE_ALL(maximum,Max);
+
+#undef SPECIALIZE_ALL
+#undef SPECIALIZE
+
+// specialize portability
+template<class Params, class BinOp>
+struct inclusive_scan<Params, BinOp, 1, false>
+{
+    using type_t = typename Params::type_t;
+    using scalar_t = typename Params::scalar_t;
+    using binop_t = typename Params::binop_t;
+    // assert T == scalar type, binop::type == T
+    using config_t = typename Params::config_t;
+
+    // affected by https://github.com/microsoft/DirectXShaderCompiler/issues/7006
+    // NBL_CONSTEXPR_STATIC_INLINE uint32_t SubgroupSizeLog2 = config_t::SizeLog2;
+
+    scalar_t operator()(scalar_t value)
+    {
+        return __call(value);
+    }
+
+    static scalar_t __call(scalar_t value)
+    {
+        binop_t op;
+        const uint32_t subgroupInvocation = glsl::gl_SubgroupInvocationID();
+
+        scalar_t rhs = glsl::subgroupShuffleUp<scalar_t>(value, 1u); // all invocations must execute the shuffle, even if we don't apply the op() to all of them
+        value = op(value, hlsl::mix(rhs, binop_t::identity, subgroupInvocation < 1u));
+
+        const uint32_t SubgroupSizeLog2 = config_t::SizeLog2;
+        [unroll]
+        for (uint32_t i = 1; i < integral_constant<uint32_t,SubgroupSizeLog2>::value; i++)
+        {
+            const uint32_t step = 1u << i;
+            rhs = glsl::subgroupShuffleUp<scalar_t>(value, step);
+            value = op(value, hlsl::mix(rhs, binop_t::identity, subgroupInvocation < step));
```

Comment:
do a quick benchmark whether `mix(op(value,rhs),value,cond)` is faster than `op(value,mix(rhs,identity,cond))`